cpp/src/rolling/lead_lag_nested_detail.cuh

```diff
@@ -0,0 +1,222 @@
+  //
+  //    Similarly, LAG(2) is implemented using the following gather_map:
+  //        {NULL_INDEX, NULL_INDEX, 0, 1, 2...}
+  //
+  // 2. Gather input column based on the gather_map.
+  // 3. If default outputs are available, scatter contents of default_outputs`
+  //    to all positions where nulls where gathered in step 2.
+  //
+  // Note: Step 3 can be switched to use `copy_if_else()`, once it supports
+  //       nested types.
+
+  auto static constexpr size_data_type = data_type{type_to_id<size_type>()};
+
+  auto gather_map_column =
+    make_numeric_column(size_data_type, input.size(), mask_state::UNALLOCATED, stream);
+  auto gather_map = gather_map_column->mutable_view();
+
+  thrust::transform(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator(size_type{0}),
+                    thrust::make_counting_iterator(size_type{input.size()}),
+                    gather_map.begin<size_type>(),
+                    lead_lag_gather_map_builder<op, PrecedingIter, FollowingIter>{
+                      input.size(), offset, preceding, following});
+
+  auto output_with_nulls = cudf::gather(table_view{std::vector<column_view>{input}},
+                                        gather_map_column->view(),
+                                        out_of_bounds_policy::NULLIFY);
+
+  if (default_outputs.is_empty()) { return std::move(output_with_nulls->release()[0]); }
+
+  // Must scatter defaults.
+  auto NULL_INDEX = size_type{input.size() + 1};
+
+  auto scatter_map = rmm::device_uvector<size_type>(input.size(), stream);
+
+  // Find all indices at which LEAD/LAG computed nulls previously.
+  auto scatter_map_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator(size_type{0}),
+                    thrust::make_counting_iterator(size_type{input.size()}),
+                    scatter_map.begin(),
+                    is_null_index_predicate(input.size(), gather_map.begin<size_type>()));
+
+  // Bail early, if all LEAD/LAG computations succeeded. No defaults need be substituted.
+  if (scatter_map.is_empty()) { return std::move(output_with_nulls->release()[0]); }
+
+  // Gather only those default values that are to be substituted.
+  auto gathered_defaults =
+    cudf::detail::gather(table_view{std::vector<column_view>{default_outputs}},
+                         scatter_map.begin(),
+                         scatter_map_end,
+                         out_of_bounds_policy::DONT_CHECK,
+                         stream);
+
+  // Scatter defaults into locations where LEAD/LAG computed nulls.
+  auto scattered_results = cudf::detail::scatter(
+    table_view{std::vector<column_view>{gathered_defaults->release()[0]->view()}},
+    scatter_map.begin(),
+    scatter_map_end,
+    table_view{std::vector<column_view>{output_with_nulls->release()[0]->view()}},
+    false,
+    stream);
+  return std::move(scattered_results->release()[0]);
```

Comment:
I don't see the `mr` parameter is used for generating the result here.

Reply:
Good catch. I'm also not using the `stream` in the first `gather()` call. I'll post a fix shortly.